man/ali.man
.TH ALI %manext1% 2013-12-05 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
ali \- list nmh mail aliases
.SH SYNOPSIS
.HP 5
.na
.B ali
.RB [ \-help ]
.RB [ \-version ]
.RB [ \-alias
.IR aliasfile ]
.RB [ \-list " | " \-nolist ]
.RB [ \-user " | " \-nouser ]
.RI [ aliases " ...]"
.ad
.SH DESCRIPTION
.B ali
searches the named mail alias files for each of the given
.IR aliases .
It creates a list of addresses for those
.IR aliases ,
and writes that list on standard output.  If no arguments are given,
.B ali
outputs all alias entries.
.PP
By default, when an alias expands to multiple addresses, the addresses
are separated by commas and printed on as few lines as possible.  If the
.B \-list
switch is specified, and an address expands to multiple
addresses, each address will appear on a separate line.
.PP
The
.B \-user
switch directs
.B ali
to perform its processing in
an inverted fashion: instead of listing the addresses that each given
alias expands to,
.B ali
will list the aliases that expand to each
given address.
.PP
The files specified by the profile entry
.RI \*(lq Aliasfile \*(rq
and any additional alias files given by the
.B \-alias
.I aliasfile
switch will be read.  Each
.I alias
is processed as described in
.IR mh\-alias (5).
.SH FILES
Alias files are looked up in multiple locations: if the pathname is
absolute or starts with
.RI \*(lq . \*(rq
or
.RI \*(lq .. \*(rq
it is accessed directly; otherwise tilde expansion is done on
usernames, then files are searched for in the user's
.I Mail
directory as specified in their profile.  If not found there, the directory
.RI \*(lq %nmhetcdir% \*(rq
is checked.
.PP
.TP 20
$HOME/.mh_profile
The user's profile.
.SH "PROFILE COMPONENTS"
.PP
.PD 0
.TP 20
Path:
To determine the user's nmh directory
.TP
Aliasfile:
For a default alias file
.PD
.SH "SEE ALSO"
.IR mh\-alias (5)
.SH DEFAULTS
.PD 0
.TP "\w'aliasfile  'u
aliasfile
%nmhetcdir%/MailAliases
.TP
\-nolist
.TP
\-nouser
.PD
.SH CONTEXT
None


NEWS
#
# NEWS - Release notes for nmh 1.8
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been nearly five years since the last release of nmh, and there have
been a number of significant changes since that last release.  Long-time
MH and nmh users should read carefully the NOTABLE CHANGES section, as
there are some significant changes to nmh behavior.  Otherwise, please
see the README and INSTALL files for help on getting started with nmh.

This release is dedicated to Norman Z. Shapiro, co-designer of the MH
Message Handling System.  MH is the predecessor of nmh.  Norm was an
active supporter of nmh development until he passed away in October of
2021.  We are most grateful to Norm for his stewardship of MH and nmh.
    https://en.wikipedia.org/wiki/Norman_Shapiro

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

- Support for Content-MD5 header fields, MIME content cache functionality,
  and the message/partial MIME type have been removed.
- Gmail OAuth2/XOAUTH support for desktop applications has been effectively
  dropped, so nmh no longer supports it.  nmh support for Gmail API access
  is experimental, please post to nmh-workers@nongnu.org if you'd like to
  help with test and development.
- repl(1) -convertargs now allows editing of the composition draft between
  translation and any encoding of text content.  Because encoding can wrap
  long lines, the use of a paragraph formatter has been removed from
  mhn.defaults.

------------
NEW FEATURES
------------

- The default editor has been changed from 'vi' to 'prompter', to align with
  historical practice and reduce packaging dependencies on external programs.
- A new -checkbase64 switch has been added to mhfixmsg(1).
- inc(1)/msgchk(1) now support STARTTLS for the POP protocol.
- All TLS-supported protocols now will send the SNI (server name indicator)
  TLS extension.
- A new mh-format function %(ordinal) has been implemented to output the
  appropriate ordinal suffix for numbers.  Example: "%(num 22)%(ordinal)"
  will output "22nd".
- show and mhl now decode more addresses in header fields.
- Added warning from all programs that read the profile if the profile
  contains a post entry, which is ignored, but does not contain a
  postproc entry.  In other words, if you get this warning and want
  to suppress it, your options include:
  1) Remove your post profile entry.
  2) Make your post profile entry a comment by prepending it with the #:
     comment indicator.
  3) Add a postproc entry that points to the post that you use.  That can
     be viewed with "mhparam postproc".
- scan(1) -file argument can be a Maildir directory.
- Updated mhn.defaults to prefer mpv(1) over xv(1) and replace mpeg_play(1),
  and to use it for all video types, not just video/mpeg.  And prefer all
  other searched-for pdf viewers over acroread(1).
- Added mhshow-suffix-video.mp4 to mhn.defaults, for use by mhshow(1) and
  send(1).
- Removed support from mhn.defaults for application/x-ivs and text/richtext.
- Changed interpretation of argument to mhfixmsg(1) -decodeheaderfieldbodies
  switch to specify character set of the decoded field bodies.
- repl(1) -convertargs now allows editing of the composition draft between
  translation and any encoding of text content.
- install-mh(1) now enables the mh-draft(5) draft folder facility.

-----------------
OBSOLETE FEATURES
-----------------

- The generation and verification of a Content-MD5 field has been removed
  without deprecation.  The related -check and -nocheck options now error.
- The MIME content cache functionality has been mostly non-functional since
  the development on nmh, and consequently all of the content caching code
  and related switches (-cache/-rcache/-wcache) have been removed.
- Support for generating and reassembling message/partial messages has been
  removed; it seems that this has been broken since 1.5 and there is very
  little support across MUAs.
- Marked Gmail OAuth2/XOAUTH support as being unsupported.
- Support for the MHPDEBUG environment variable was removed.  It was
  deprecated in nmh 1.7.  The pick(1) -debug switch replaced it.
- The 'libdir' mhparam(1) component was removed.  It was deprecated in
  nmh 1.7, when it was replaced by a new 'libexecdir' component.

---------
BUG FIXES
---------

- Fixed bcc to work with sendmail/pipe, and better documented that dcc
  doesn't work with it [Bug 55700].
- An -attendee switch has been added to mhical(1), for use when more than one
  (or zero) attendees match a user's mailbox.
- Fixed inc(1) and %(me) function escape to not obey Local-Mailbox profile
  component.
- Fixed source charset in mhfixmsg textcharset verbose output.
- Fixed mhfixmsg charset determination of content added with -reformat.
- Fixed file descriptor leak in mhfixmsg when run on multiple input files.
- Fixed mhfixmsg(1) -decodeheaderfilebodies to support mixed encoded/undecoded.
- Fixed memory corruption in post(1) and inc(1) when using XOAUTH2,
  with 4 or more entries in the oauth-authservice file.
- Added alias expansion to From: address for use by sendfrom.
- Removed extra space added before header field bodies by dist(1) to $mhdraft.
- Fixed display of iCalendar object with multiple VEVENTS using a VTIMEZONE.
- Fixed allowable encodings with MIME message types to get closer to RFC 2046.
- Detect other files regardless of backup prefix [Bug #49476].
- Copy if hard link by refile(1) or send(1) fails with EACESS [Bug #56575].

----------------------------------------------------------------------------

#
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time

uip/ali.c
/* ali.c -- list nmh mail aliases
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include "h/mh.h"
#include "sbr/getarguments.h"
#include "sbr/smatch.h"
#include "sbr/getcpy.h"
#include "sbr/context_find.h"
#include "sbr/brkstring.h"
#include "sbr/ambigsw.h"
#include "sbr/print_version.h"
#include "sbr/print_help.h"
#include "sbr/error.h"
#include "sbr/addrsbr.h"
#include "aliasbr.h"
#include "sbr/mts.h"
#include "sbr/done.h"
#include "sbr/utils.h"
#include "sbr/globals.h"

#define ALI_SWITCHES \
    X("alias aliasfile", 0, ALIASW) \
    X("noalias", 0, NALIASW) \
    X("list", 0, LISTSW) \
    X("nolist", 0, NLISTSW) \
    X("user", 0, USERSW) \
    X("nouser", 0, NUSERSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(ALI);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(ALI, switches);
#undef X

static int pos = 1;

/*
 * prototypes
 */
static void print_aka (char *, bool, int);
static void print_usr (char *, bool);


int
main (int argc, char **argv)
{
    int i, vecp = 0;
    bool inverted, list, noalias;
    char *cp, **ap, **argp, buf[BUFSIZ];
    /* Really only need to allocate for argc-1, but must allocate at least 1,
       so go ahead and allocate for argc char pointers. */
    char **vec = mh_xmalloc (argc * sizeof (char *)), **arguments;
    struct aka *ak;

    if (nmh_init(argv[0], true, true)) { return 1; }

    mts_init ();
    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    inverted = list = noalias = false;
    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    die("-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [switches] aliases ...",
			invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version (invo_name);
		    done (0);

		case ALIASW: 
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    if ((i = alias (cp)) != AK_OK)
			die("aliasing error in %s - %s", cp, akerror (i));
		    continue;
		case NALIASW: 
		    noalias = true;
		    continue;

		case LISTSW: 
		    list = true;
		    continue;
		case NLISTSW: 
		    list = false;
		    continue;

		case USERSW: 
		    inverted = true;
		    continue;
		case NUSERSW: 
		    inverted = false;
		    continue;
	    }
	}

	if (vecp < argc) {
	    vec[vecp++] = cp;
	} else {
	    /* Should never happen, but try to protect against code changes
	       that could allow it. */
	    die("too many arguments");
	}
    }

    if (!noalias) {
	/* allow Aliasfile: profile entry */
	if ((cp = context_find ("Aliasfile"))) {
	    char *dp = NULL;

	    for (ap = brkstring(dp = mh_xstrdup(cp), " ", "\n"); ap && *ap; ap++)
		if ((i = alias (*ap)) != AK_OK)
		    die("aliasing error in %s - %s", *ap, akerror (i));
            free(dp);
	}
	alias (AliasFile);
    }

    /*
     * If -user is specified
     */
    if (inverted) {
	if (vecp == 0)
	    die("usage: %s -user addresses ...  (you forgot the addresses)",
		   invo_name);

	for (i = 0; i < vecp; i++)
	    print_usr (vec[i], list);
    } else {
	if (vecp) {
	    /* print specified aliases */
	    for (i = 0; i < vecp; i++)
		print_aka (akvalue (vec[i]), list, 0);
	} else {
	    /* print them all */
	    for (ak = akahead; ak; ak = ak->ak_next) {
                char *res;

		printf ("%s: ", ak->ak_name);
		pos += strlen (ak->ak_name) + 1;
                res = akresult(ak);
		print_aka(res, list, pos);
                free(res);
	    }
	}
    }

    free (vec);
    done (0);
    return 1;
}

static void
print_aka (char *p, bool list, int margin)
{
    char c;

    if (p == NULL) {
	puts("<empty>");
	return;
    }

    while ((c = *p++)) {
	switch (c) {
	    case ',': 
		if (*p) {
		    if (list)
			printf ("\n%*s", margin, "");
		    else {
			if (pos >= 68) {
			    fputs(",\n ", stdout);
			    pos = 2;
			} else {
			    fputs(", ", stdout);
			    pos += 2;
			}
		    }
		}

	    case 0: 
		break;

	    default: 
		pos++;
		putchar (c);
	}
    }

    putchar ('\n');
    pos = 1;
}

static void
print_usr (char *s, bool list)
{
    char *cp, *pp, *vp;
    struct aka *ak;
    struct mailname *mp, *np;

    if ((pp = getname (s)) == NULL)
	die("no address in \"%s\"", s);
    if ((mp = getm (pp, NULL, 0, NULL, 0)) == NULL)
	die("bad address \"%s\"", s);
    while (getname (""))
	continue;

    vp = NULL;
    for (ak = akahead; ak; ak = ak->ak_next) {
	pp = akresult (ak);
	while ((cp = getname (pp))) {
	    if ((np = getm (cp, NULL, 0, NULL, 0)) == NULL)
		continue;
	    if (!strcasecmp (FENDNULL(mp->m_host),
			     FENDNULL(np->m_host))  &&
		!strcasecmp (FENDNULL(mp->m_mbox),
			     FENDNULL(np->m_mbox))) {
		vp = vp ? add (ak->ak_name, add (",", vp))
		    : getcpy (ak->ak_name);
		mnfree (np);
		while (getname (""))
		    continue;
		break;
	    }
	    mnfree (np);
	}
    }
    mnfree (mp);

    print_aka (vp ? vp : s, list, 0);

    free(vp);
}
